Add blog post on Native Java with Helidon (#985)
oktadeveloper/okta-blog: _source/_posts/2021-12-18-native-java-helidon.adoc

---
layout: blog_post
title: "Build REST APIs and Native Java Apps with Helidon"
author: matt-raible
by: advocate
communities: [java]
description: "Learn how to create REST APIs and native images with Helidon, a Java microservices framework from Oracle."
tags: [native, java, graalvm, helidon]
tweets:
- ""
- ""
- ""
image:
type: conversion
github: https://github.com/oktadev/native-java-examples
---
:page-liquid:
:toc: macro
:experimental:

> Project Helidon is a set of Java libraries for writing microservices. It's an open source, Apache 2.0-licensed, Oracle-sponsored project and contains support for MicroProfile as well as a reactive, functional API. Helidon's two different style are known as Helidon MP and SE, respectively. The https://helidon.io[helidon.io] website says:

Reviewer: ```suggestion
Project Helidon is a set of Java libraries for writing microservices. It's an open source, Apache 2.0-licensed, Oracle-sponsored project and contains support for MicroProfile as well as a reactive, functional API. Helidon's two different styles are known as Helidon MP and SE, respectively. The https://helidon.io[helidon.io] website says:
```
plural

Author: I don't know why, but GitHub has the "Commit suggestion" button grayed out.

Author: I think this is because this suggestion is on a commit, not on the PR. I'll fix manually.

====

> Since Helidon is simply a collection of Java libraries running on a fast Netty core, there is no extra overhead or bloat.

Reviewer: Does this need quotation marks? 
A link to Netty (https://netty.io/)?

Author: I used the ==== to indicate a quotation. If you look at the preview, you'll see it surrounds the text with a yellow box. I could use > as well, but I don't like the look of that. I don't think linking to Netty is relevant since it's not linked to by Helidon on their site.

====

image::{% asset_path 'blog/native-java-helidon/helidon-mp.png' %}[alt=Helidon supports MicroProfile,width=400,align=center]

> Today, I'm going to put this statement to the test! First, I'll show you how to run a secure, OAuth 2.0-protected, Java REST API that allows JWT authentication. Then, I'll show you how to build a native image with GraalVM. Finally, I'll compare Helidon to the most popular Java REST API frameworks: Micronaut, Quarkus, and Spring Book.

Reviewer: ```suggestion
Today, I'm going to put this statement to the test! First, I'll show you how to run a secure, OAuth 2.0-protected Java REST API that allows JWT authentication. Then, I'll show you how to build a native image with GraalVM. Finally, I'll compare Helidon to the most popular Java REST API frameworks: Micronaut, Quarkus, and Spring Book.
```
remove comma because "OAuth 2.0-protected" modifies Java REST API directly

My link:/blog/2021/06/18/native-java-framework-comparison[Build Native Java Apps with Micronaut, Quarkus, and Spring Boot] post compares startup times and memory usage of these three frameworks. How does Helidon perform? Read on!

**Prerequisites**

- https://sdkman.io/[SDKMAN]
- https://maven.apache.org[Maven]
- https://httpie.io/[HTTPie] (a better version of cURL)
- An https://developer.okta.com[Okta Developer] Account

toc::[]

> == Get Started with Helidon

Reviewer: ```suggestion
== Get started with Helidon
```
sentence case

I created a GitHub repository you can clone and run to get started with Helidon quickly.

[source,shell]
----

> git clone https://github.com/oktadev/native-java-examples.git

Reviewer 2: Don't forget to merge your `helidon` branch 😄

Author: This will be merged, so no branch reference necessary.

cd native-java-examples/helidon
----

> I'll show you how I created this example in a section below. First, let's see how to run it.

Reviewer: ```suggestion
In a section below, I'll show you how I created this example. First, let's see how to run it.
```
Removes ambiguity

Open the `helidon` directory in your favorite IDE, so you have easy access to the example's project files.

TIP: If you just want to see how to build a native image, skip to the link:#build-native-images-for-helidon[build native images for Helidon] section.

== Install a JDK with GraalVM

You will need a JDK with GraalVM and its `native-image` compiler. Using SDKMAN, run the following command and set it as the default:

[source,shell]
----

> sdk install java 21.3.0.r17-grl

Reviewer 2: Nice, 17!

----

Add the native extension to the JDK:

[source,shell]
----
gu install native-image
----

== Run a Helidon Java API

In a terminal window, run `mvn package` to package the app into a JAR. Then use `java -jar` to run it.

[source,shell]
----
mvn package

> java -jar target/helidon.jar

Reviewer 2: You can also run `mvn exec:exec`
or set the default goal in your project pom: `<defaultGoal>exec:exec</defaultGoal>`

Pet peeve, one of the Helidon parent poms uses an older version of the `exec-maven-plugin`, ideally, this _would_ use `exec:java` instead, but I'm splitting hairs, and they probably had a reason for it 😄 

I was also able to run and debug the app through IntelliJ by creating a run configuration with `io.helidon.microprofile.cdi.Main`

Author: I don't have an opportunity to try this right now. Do you think I should add it or just leave it out?

----

> TIP: Helidon's Maven Plugin doesn't have a _run_ command. You can install the https://helidon.io/docs/v2/#/about/05_cli[Helidon CLI] and run `helidon dev`. Running this command allows you to change code and automatically recompile + restart your application.

Reviewer: ```suggestion
TIP: Helidon's Maven Plugin doesn't have a _run_ command. You can install the https://helidon.io/docs/v2/#/about/05_cli[Helidon CLI] and run `helidon dev`. Running this command allows you to change code and automatically recompile and restart your application.
```
replace "+"

If you open another terminal window and try to access it with HTTPie, you'll get a 401 Unauthorized error.

[source,shell]
----
$ http :8080/hello

HTTP/1.1 401 Unauthorized
Content-Length: 0
Date: Fri, 3 Dec 2021 06:11:25 -0700
connection: keep-alive
----

To make it so you can access this endpoint, you'll need to generate an OAuth 2.0 access token and update the JWKS (JSON Web Key Sets) URL to yours (in this project's `application.yml`).

NOTE: If you're unsure what OIDC and OAuth 2.0 are, see link:/blog/2017/06/21/what-the-heck-is-oauth[What the Heck is OAuth?].

{% include setup/cli.md type="spa" loginRedirectUri="https://oidcdebugger.com/debug" logoutRedirectUri="https://oidcdebugger.com" %}

Take note of the `clientId` and `issuer` values. You'll need those to get an access token and to configure each framework for JWT authentication.

Open `src/main/resources/META-INF/microprofile-config.properties` and change the Okta URL to match yours.

[source,properties]
----
mp.jwt.verify.publickey.location=https://{yourOktaDomain}/oauth2/default/v1/keys

> mp.jwt.verify.issuer=https://{yourOktaDomain}/oauth2/default

Reviewer 2: nit: `https://{yourOktaDomain}/oauth2/default` should probably be {issuer} in all places (as that is what the CLI outputs)

Reviewer 2: That said, we use `yourOktaDomain` all over the place, so that is fairly consistent too, so feel free to ignore this

> mp.jwt.verify.publickey.location=https://{yourOktaDomain}/oauth2/default/v1/keys
> mp.jwt.verify.issuer=https://{yourOktaDomain}/oauth2/default

Reviewer 2: ```suggestion
mp.jwt.verify.issuer={issuer}
mp.jwt.verify.publickey.location=${mp.jwt.verify.issuer}/v1/keys
```
DRY out the configuration

Author: Did you test this to ensure it works? I don't have a laptop available to test, that's why I ask.

Author: Just tested, it works.

----

Stop your Helidon app with kbd:[Ctrl + C], repackage it, and re-run it with kbd:[⬆️ + Return].

[source,shell]
----
mvn package
java -jar target/helidon.jar
----

> === Generate an OAuth 2.0 Access Token

Reviewer: ```suggestion
=== Generate an OAuth 2.0 access token
```
sentence case

> {% include setup/oidcdebugger.md %}

Reviewer 2: We should try to get someone to add PKCE support to the OIDC debugger, it would trim off a few steps

Author: I added a comment to see if we could sponsor this feature. https://github.com/nbarbettini/oidc-debugger/issues/39

image::{% asset_path 'blog/native-java-comparison/oidc-debugger.png' %}[alt=OIDC Debugger,width=600,align=center]

Click **Send Request** to continue.

Once you have an access token, set it as a `TOKEN` environment variable in a terminal window.

[source,shell]
----
TOKEN=eyJraWQiOiJZMVRxUkRQbEFEcm1XN0dX...
----

> === Test Your Helidon API with HTTPie

Reviewer: ```suggestion
=== Test your Helidon API with HTTPie
```
sentence case

Use HTTPie to pass the JWT in as a bearer token in the `Authorization` header.

[source,shell]
----
http :8080/hello Authorization:"Bearer $TOKEN"
----

> You should get a 200 response with your email in it.

Reviewer: ```suggestion
You should get a 200 response with your email address  in it.
```
clarity

image::{% asset_path 'blog/native-java-helidon/httpie-helidon-bearer-token.jpg' %}[alt=HTTPie call to Helidon's /hello with bearer token,width=800,align=center]

> === Build a Native Helidon App

Reviewer: ```suggestion
=== Build a native Helidon app
```
sentence case

To compile this Micronaut app into a native binary, run:

[source,shell]
----
mvn package -Pnative-image
----

This command will take a few minutes to complete. My 2019 MacBook Pro with a 2.4 GHz 8-Core Intel Core i9 processor and 64 GB of RAM took 2 min. 34 s. to finish.

Start it with `./target/helidon`:

----
$ ./target/helidon
...

2021.12.03 06:18:13 INFO io.helidon.microprofile.server.ServerCdiExtension Thread[main,5,main]: Server started on http://localhost:8080 (and all other host addresses) in 550 milliseconds (since JVM startup).
2021.12.03 06:18:13 INFO io.helidon.common.HelidonFeatures Thread[features-thread,5,main]: Helidon MP 2.4.0 features: [CDI, Config, Fault Tolerance, Health, JAX-RS, Metrics, Open API, REST Client, Security, Server, Tracing, Web Client]
2021.12.03 06:18:13 INFO io.helidon.common.HelidonFeatures.experimental Thread[features-thread,5,main]: You are using experimental features. These APIs may change, please follow changelog!
2021.12.03 06:18:13 INFO io.helidon.common.HelidonFeatures.experimental Thread[features-thread,5,main]: 	Experimental feature: Web Client (WebClient)
----

> You can see it starts in a just over half a second (550ms)! Test it with HTTPie and an access token. You may have to generate a new JWT with https://oidcdebugger.com[oidcdebugger.com] if yours has expired.

Reviewer: ```suggestion
As you can see, it starts in a just over half a second (550 milleseconds)! Test it with HTTPie and an access token. You may have to generate a new JWT with https://oidcdebugger.com[oidcdebugger.com] if yours has expired.
```
clarity

[source,shell]
----
http :8080/hello Authorization:"Bearer $TOKEN"
----

> === Create a Helidon App from Scratch

Reviewer: ```suggestion
=== Create a Helidon app from scratch
```
sentence case

> You might be wondering, "how did you build a secure Micronaut app"? Did I just hide the complexity? No, it only takes five steps to create the same app.

Reviewer: ```suggestion
You might be wondering, "how did you build a secure Micronaut app"? Did I just hide the complexity? No. It only takes five steps to create the same app.
```
clarity

    -DgroupId=com.okta.rest \
    -DartifactId=helidon \
    -Dpackage=com.okta.rest
+
If you use `helidon init`, answer the questions like I did.
+
image::{% asset_path 'blog/native-java-helidon/helidon-init.jpg' %}[alt=helidon init command,width=800,align=center]

. Add a `HelloResource` in `src/main/java/com/okta/rest/controller/HelloResource.java`:
+
[source,java]
----
package com.okta.rest.controller;

import io.helidon.security.Principal;
import io.helidon.security.SecurityContext;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;
import java.util.Optional;

import static javax.ws.rs.core.MediaType.TEXT_PLAIN;

@Path("/hello")
public class HelloResource {

    @GET
    @Path("/")
    @Produces(TEXT_PLAIN)
    public String hello(@Context SecurityContext context) {
        Optional<Principal> userPrincipal = context.userPrincipal();
        return "Hello, " + userPrincipal.get().getName() + "!";
    }
}
----

. Enable and configure JWT security in `src/main/resources/META-INF/microprofile-config.properties`:
+
[source,properties]
----
mp.jwt.verify.publickey.location=https://{yourOktaDomain}/oauth2/default/v1/keys

> mp.jwt.verify.issuer=https://{yourOktaDomain}/oauth2/default

Reviewer 2: See note above

----

. Add a `HelloApplication` class in `src/main/java/com/okta/rest` to register your resource and configure JWT authentication:
+
[source,java]
----
package com.okta.rest;

import com.okta.rest.controller.HelloResource;
import org.eclipse.microprofile.auth.LoginConfig;

import javax.enterprise.context.ApplicationScoped;
import javax.ws.rs.core.Application;
import java.util.Set;

@LoginConfig(authMethod = "MP-JWT")
@ApplicationScoped
public class HelloApplication extends Application {
    @Override
    public Set<Class<?>> getClasses() {
        return Set.of(HelloResource.class);
    }
}
----

. Delete `GreetingsProvider`, `GreetResource`, and `MainTest` since they're not used in this example.

  rm src/main/java/com/okta/rest/Greet*
  rm src/test/java/com/okta/rest/MainTest.java

That's it! Now you can start the app or build the native image as shown above.

> == Native Java Startup Time Comparison

Reviewer: ```suggestion
== Native Java startup time comparison
```
sentence case

To compare startup times between Micronaut, Quarkus, Spring Boot, and Helidon, I first created native executables. You can run the following commands to do the same from the root directory of the example you cloned.

[source,shell]
----
cd micronaut
./mvnw package -Dpackaging=native-image

cd ../quarkus
./mvnw package -Pnative

cd ../spring-boot
./mvnw package -Pnative

# Helidon should already be built, but just in case it isn't
cd ../helidon
mvn package -Pnative-image
----

I ran each image three times before I started recording the numbers. I then ran each command five times.

_These numbers are from a 2019 MacBook Pro with a 2.4 GHz 8-Core Intel Core i9 processor and 64 GB of RAM. I think it's important to note that my WiFi connection was 71 Mbps down and 10 Mbps up (according to the Speedtest app)._

////
Helidon: (288 + 495 + 280 + 291 + 540) / 5 = 378.8
Micronaut: (20 + 19 + 21 + 22 + 20) / 5 = 20.4
Quarkus: (18 + 16 + 16 + 18 + 17) / 5 = 17
Spring Boot: (59 + 59 + 58 + 59 + 61) / 5 = 59.2
////

> .Native Java Startup times in milliseconds

Reviewer: ```suggestion
.Native Java startup times in milliseconds
```
capitalization not needed

|===
|Framework | Command executed | Milliseconds to start

|Helidon | `./helidon/target/helidon` | 378.8
|Micronaut | `./micronaut/target/app` | 20.4
|Quarkus | `./quarkus/target/quarkus-1.0.0-SNAPSHOT-runner` | 17
|Spring Boot | `./spring-boot/target/demo` | 59.2
|===

The chart below should help you visualize this comparison.

++++
<script src="https://www.gstatic.com/charts/loader.js"></script>
<div id="startup-times"></div>
<script>
google.charts.load('current', {packages: ['corechart', 'bar']});
google.charts.setOnLoadCallback(drawChart);
function drawChart() {
  var data = google.visualization.arrayToDataTable([
    ['Framework', 'Milliseconds to start', { role: 'style' }],
    ['Helidon', 378.8, 'orange'],
    ['Micronaut', 20.4, 'blue'],
    ['Quarkus', 17, 'red'],
    ['Spring Boot', 59.2, 'green']
  ]);
  var options = {

>     title: 'Startup times of Native Java frameworks',

Reviewer: ```suggestion
    title: 'Startup times of native Java frameworks',
```
capitalization

----

I substituted `<executable>` as follows:

.Native Java memory used in megabytes
|===
|Framework | Executable | Megabytes before request | Megabytes after request

|Helidon | `helidon` | 74 | 110
|Micronaut | `app` | 27 | 58
|Quarkus | `runner` | 20 | 32
|Spring Boot | `demo` | 48 | 62
|===

++++
<div id="memory-usage"></div>
<script>
google.charts.load('current', {packages: ['corechart', 'bar']});
google.charts.setOnLoadCallback(drawChart);
function drawChart() {
  var data = google.visualization.arrayToDataTable([
    ['Framework', 'Memory usage (MB)', { role: 'style' }],
    ['Helidon', 110, 'orange'],
    ['Micronaut', 58, 'blue'],
    ['Quarkus', 32, 'red'],
    ['Spring Boot', 62, 'green']
  ]);
  var options = {
    title: 'Memory usage of Native Java frameworks',
    chartArea: {width: '50%'},
    hAxis: {
      title: 'Megabytes',
      minValue: 0
    },
    vAxis: {
      title: 'Java Framework'
    }
  };
  var chart = new google.visualization.BarChart(document.getElementById('memory-usage'));
  chart.draw(data, options);
}
</script>
++++

This data shows that Helidon has some work to do to catch up to its competition!

> == Learn More About Java and GraalVM

Reviewer: ```suggestion
== Learn more about Java and GraalVM
```

> In this post, you learned how to develop, build, and run native a Java app with Helidon. You also learned how to secure it with OpenID Connect and access it with a JWT access token.

Reviewer: ```suggestion
In this post, you learned how to develop, build, and run a native Java app with Helidon. You also learned how to secure it with OpenID Connect and access it with a JWT access token.
```
typo
